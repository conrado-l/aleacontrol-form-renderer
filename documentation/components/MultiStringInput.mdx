---
name: MultiStringInput
---

import { Playground, PropsTable } from 'docz'
import MultiStringInput from '../../src/components/MultiStringInput.js'

# MultiStringInput

<PropsTable of={MultiStringInput} />

## Description
Multiple string inputs with the ability to add and remove them individually
## Demo and usage

<Playground>
    <MultiStringInput value={['MSi', 'Intel', 'AMD']} name='brands' col={4} update={(e) => e}/>
</Playground>
