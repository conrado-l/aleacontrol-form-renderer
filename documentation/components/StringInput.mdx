---
name: StringInput
---

import { Playground, PropsTable } from 'docz'
import StringInput from '../../src/components/StringInput.js'

# StringInput

<PropsTable of={StringInput} />

## Description
A simple text input
## Demo and usage

<Playground>
    <StringInput value='Hello' name='greeting' col={4} update={(e) => e}/>
</Playground>
