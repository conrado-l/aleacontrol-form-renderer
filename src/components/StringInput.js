import React, { Component } from 'react'
import PropTypes from 'prop-types'
import Row from 'react-bootstrap/Row'
import Col from 'react-bootstrap/Col'
import Form from 'react-bootstrap/Form'

/**
 * Regular text type input
 */
class StringInput extends Component {
  constructor (props) {
    super(props)
    this.state = {
      value: props.value
    }

    this.handleChange = this.handleChange.bind(this)
  }

  /**
   * Updates the input value and sends the value through the parent callback function
   * @param {object} e Event
   */
  handleChange (e) {
    const newValue = e.target.value
    this.setState({ value: newValue })
    this.props.update(this.props.name, newValue)
  }

  render () {
    return (
      <Row>
        <Col xs={12} sm={this.props.col}>
          <Form.Control
            type='text'
            name={this.props.name}
            value={this.state.value || ''}
            onChange={this.handleChange} />
        </Col>
      </Row>
    )
  }
}

StringInput.propTypes = {
  value: PropTypes.string,
  name: PropTypes.string,
  col: PropTypes.number,
  update: PropTypes.func
}

export default StringInput
