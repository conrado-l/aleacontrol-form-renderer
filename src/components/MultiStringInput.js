import React, { Component } from 'react'
import PropTypes from 'prop-types'
import StringInput from './StringInput'
import Button from 'react-bootstrap/Button'

/**
 * Multiple and different string inputs with the ability to add or delete them individually
 */
class MultiStringInput extends Component {
  constructor (props) {
    super(props)

    this.state = {
      stringInputs: props.value
    }

    this.addInput = this.addInput.bind(this)
    this.updateInput = this.updateInput.bind(this)
    this.deleteLastInput = this.deleteLastInput.bind(this)
  }

  /**
   * Send the inputs to the parent and filter the empty inputs
   */
  updateMultiString () {
    this.props.update(this.props.name, this.state.stringInputs.filter((input) => input !== ''))
  }

  /**
  * Adds a new input
  */
  addInput () {
    this.setState(prevState => ({
      stringInputs: [...prevState.stringInputs, '']
    }), () => this.updateMultiString())
  }

  /**
   * Deletes the last input
   */
  deleteLastInput () {
    let deletedLastInputState = [...this.state.stringInputs]
    deletedLastInputState.splice(this.state.stringInputs.length - 1, 1)
    this.setState({ stringInputs: deletedLastInputState }, () => this.updateMultiString())
  }

  /**
  * Updates an input value
  */
  updateInput (index, value) {
    let inputs = this.state.stringInputs
    inputs[index] = value
    this.setState({
      stringInputs: inputs
    }, () => this.updateMultiString())
  }

  /**
   * Renders the inputs
   */
  renderInputs () {
    return this.state.stringInputs.map((input, index) => {
      return <div className='mb-3' key={index}>
        <StringInput name={`${index}`} value={input} col={this.props.col} update={this.updateInput} />
      </div>
    })
  }

  render () {
    return (
      <div>
        {this.renderInputs()}
        <div className='d-flex justify-content-around'>
          <Button id='add' variant='outline-success' onClick={this.addInput}> Add input</Button>
          <Button id='delete' variant='outline-danger' disabled={!this.state.stringInputs.length} onClick={this.deleteLastInput}> Delete last input </Button>
        </div>
      </div>
    )
  }
}

MultiStringInput.propTypes = {
  name: PropTypes.string,
  value: PropTypes.array,
  col: PropTypes.number,
  update: PropTypes.func
}

export default MultiStringInput
