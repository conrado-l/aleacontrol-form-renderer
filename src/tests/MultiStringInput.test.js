import React from 'react'
import { shallow, mount } from 'enzyme'
import MultiStringInput from '../components/MultiStringInput'
import StringInput from '../components/StringInput'
import renderer from 'react-test-renderer'

describe('MultiStringInput', () => {
  const mockupProps = {
    name: 'interests',
    value: ['Running', 'Play the piano', 'Coding']
  }

  it('should render correctly multistring component', () => {
    const MultiStringInputComponent = renderer
      .create(<MultiStringInput name={mockupProps.name} value={mockupProps.value} />)
      .toJSON()
    expect(MultiStringInputComponent).toMatchSnapshot()
  })

  it('should render N string inputs with the proper value', () => {
    const MultiStringInputComponent = mount(<MultiStringInput name={mockupProps.name} value={mockupProps.value} />)
    expect(MultiStringInputComponent.find(StringInput)).toHaveLength(mockupProps.value.length)

    const texts = MultiStringInputComponent.find('input').map(node => node.instance().value)
    expect(texts).toEqual(mockupProps.value)
  })

  it('should update the stringInputs state when the inputs update', () => {
    const updateFn = jest.fn()
    const MultiStringInputComponent = mount(<MultiStringInput name={mockupProps.name} value={mockupProps.value} update={updateFn} />)

    MultiStringInputComponent.find('input').map((input, index) => {
      input.simulate('change', {
        target: {
          value: index.toString()
        }
      })
    })

    expect(MultiStringInputComponent.state().stringInputs).toEqual(['0', '1', '2'])
  })

  it('should emit the local state to the parent through the update callback', () => {
    const updateFn = jest.fn()
    const MultiStringInputComponent = shallow(<MultiStringInput name={mockupProps.name} value={mockupProps.value} update={updateFn} />)
    const MultiStringInputInstance = MultiStringInputComponent.instance()

    MultiStringInputInstance.updateMultiString(mockupProps.name, mockupProps.value)

    expect(updateFn).toHaveBeenCalledWith(mockupProps.name, mockupProps.value)
  })

  it('should delete the last input and update the state', () => {
    const updateFn = jest.fn()
    const MultiStringInputComponent = mount(<MultiStringInput name={mockupProps.name} value={mockupProps.value} update={updateFn} />)
    const mockUpdatedInputs = mockupProps.value.slice(0, -1)

    MultiStringInputComponent.find('button#delete').simulate('click')

    expect(MultiStringInputComponent.find(StringInput)).toHaveLength(2)
    expect(MultiStringInputComponent.state().stringInputs).toEqual(mockUpdatedInputs)
  })

  it('should add a new empty input and update the state', () => {
    const updateFn = jest.fn()
    const MultiStringInputComponent = mount(<MultiStringInput name={mockupProps.name} value={mockupProps.value} update={updateFn} />)
    const mockUpdatedInputs = [...mockupProps.value, '']

    MultiStringInputComponent.find('button#add').simulate('click')

    expect(MultiStringInputComponent.find(StringInput)).toHaveLength(4)
    expect(MultiStringInputComponent.find('input').at(3).instance().value).toBe('')
    expect(MultiStringInputComponent.state().stringInputs).toEqual(mockUpdatedInputs)
  })
})
