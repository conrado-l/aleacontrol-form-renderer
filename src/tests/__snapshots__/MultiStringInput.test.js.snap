// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MultiStringInput should render correctly multistring component 1`] = `
<div>
  <div
    className="mb-3"
  >
    <div
      className="row"
    >
      <div
        className="col-12"
      >
        <input
          className="form-control"
          name="0"
          onChange={[Function]}
          type="text"
          value="Running"
        />
      </div>
    </div>
  </div>
  <div
    className="mb-3"
  >
    <div
      className="row"
    >
      <div
        className="col-12"
      >
        <input
          className="form-control"
          name="1"
          onChange={[Function]}
          type="text"
          value="Play the piano"
        />
      </div>
    </div>
  </div>
  <div
    className="mb-3"
  >
    <div
      className="row"
    >
      <div
        className="col-12"
      >
        <input
          className="form-control"
          name="2"
          onChange={[Function]}
          type="text"
          value="Coding"
        />
      </div>
    </div>
  </div>
  <div
    className="d-flex justify-content-around"
  >
    <button
      className="btn btn-outline-success"
      disabled={false}
      id="add"
      onClick={[Function]}
      type="button"
    >
       Add input
    </button>
    <button
      className="btn btn-outline-danger"
      disabled={false}
      id="delete"
      onClick={[Function]}
      type="button"
    >
       Delete last input 
    </button>
  </div>
</div>
`;
